### ISSUE TITLE
Roadmap

### ISSUE THREAD

**Issue by MatteoRobbiati** [qiboteam/qiboml#44]
Next steps with priorities ( High priority $\rightarrow$ 🔴 🟠 🟢 $\leftarrow$ low priority ):

### Core developing

- [x] merge `model` branch into main (🔴);
- [x] clean up the main from old/unused files (🔴);
- [x] merge PSR checking gradients (🔴);
- [ ] speed up tests (🔴):
    - [ ]  reduce the tested combinations (🔴);
    - [ ] setup a wise test procedure (🟠);
- [x] setup the differentiation procedure for data $x$ in the differentiation rules;
- [x] enable a single `BackPropagation(DifferentiationRule)` node in our differentiation pipeline (🟠) ?? - open discussion here; 
- [x] Keras interface (🟠);
- [ ] Jax backend improvements (🟠);
- [ ] Implement reuploading encoding and make the model more flexible to support trainable encoding (🟠);
    - [x] implement a first working example (🟠);
    - [ ] define how the QuantumModel structure should get as argument a trainable layer/a quantum circuit, as pointed out in comment https://github.com/qiboteam/qiboml/pull/42#discussion_r1829029612 (🟠).      
- [ ] Jax (Flax) interface (🟢); 
- [x] test GPUs work with Pytorch, Tensorflow and Jax (🟢);
- [ ] support batch execution (🟢);
- [ ] real time error mitigation (🟢);
- [ ] support a set of hardware-compatible optimizers (🟢). Have a look to [this repo of gradient-free optimizers](https://github.com/SimonBlanke/Gradient-Free-Optimizers);
- [ ] Add a function to draw the QuantumModel (just like Circuit.draw()) (🟢);

### Support tasks

- [x] migrate the pytorch backend from Qibo to Qiboml (🔴);
- [x] merge docs draft (🔴);
- [x] docs (🔴):
    - [x] webpage (🟠);
    - [x] tutorial (🔴);
- [ ] move variational examples from Qibo to Qiboml with tests (using Qiboml API) (🟠);
